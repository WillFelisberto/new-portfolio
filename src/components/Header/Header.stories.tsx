import { Meta, StoryObj } from '@storybook/react';
import Header from '.';

const meta: Meta<typeof Header> = {
  title: 'Components/Header',
  component: Header,
  args: {
    navItems: [
      { label: 'Home', href: '/' },
      { label: 'About', href: '/about' },
      { label: 'Contact', href: '/contact' },
    ],
  },
  parameters: {
    layout: 'fullscreen', // Mostra o componente ocupando toda a largura
  },
} satisfies Meta<typeof Header>;

export default meta;

type Story = StoryObj<typeof Header>;

export const Default: Story = {
  args: {
    navItems: [
      { label: 'Home', href: '/' },
      { label: 'About', href: '/about' },
      { label: 'Contact', href: '/contact' },
    ],
  },
  render: (args) => <Header {...args} />,
};

export const CustomNavItems: Story = {
  args: {
    navItems: [
      { label: 'Dashboard', href: '/dashboard' },
      { label: 'Settings', href: '/settings' },
      { label: 'Profile', href: '/profile' },
    ],
  },
  render: (args) => <Header {...args} />,
};
