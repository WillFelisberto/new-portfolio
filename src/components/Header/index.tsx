'use client';

import Image from 'next/image';
import Link from 'next/link';
import { motion } from 'framer-motion';
import { GoogleAnalytics } from './google-analytics';
import NavItem from './nav-item';

type NavItemType = {
  label: string;
  href: string;
};

export type HeaderProps = {
  navItems?: NavItemType[];
};

const NAV_ITEMS: NavItemType[] = [
  {
    label: 'Home',
    href: '/',
  },
  {
    label: 'Projects',
    href: '/projects',
  },
];

const Header = ({ navItems = NAV_ITEMS }: HeaderProps) => {
  return (
    <>
      <GoogleAnalytics />
      <motion.header
        className="absolute top-0 z-10 flex h-24 w-full items-center justify-center"
        initial={{ top: -100 }}
        animate={{ top: 0 }}
        transition={{ duration: 0.5 }}
      >
        <div className="container flex items-center justify-between">
          <Link href="/">
            <Image width={58} height={49} src="/images/logo.svg" alt="Logo" />
          </Link>
          <nav className="flex items-center gap-4 sm:gap-10">
            {navItems.map((item: NavItemType) => (
              <NavItem {...item} key={item.label} />
            ))}
          </nav>
        </div>
      </motion.header>
    </>
  );
};

export default Header;
