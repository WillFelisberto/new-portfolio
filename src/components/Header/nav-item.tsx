'use client';

import Link from 'next/link';
import { usePathname } from 'next/navigation';
import { cn } from '@/lib/utils';

type NavItemProps = {
  label: string;
  href: string;
};
const NavItem = ({ label, href }: NavItemProps) => {
  const pathname = usePathname();
  const isActive = pathname === href;

  return (
    <Link
      className={cn(
        'flex items-center gap-2 font-mono font-medium text-gray-400',
        isActive && 'text-gray-50',
      )}
      href={href}
    >
      <span className="text-sky-400">#</span>
      {label}
    </Link>
  );
};

export default NavItem;
