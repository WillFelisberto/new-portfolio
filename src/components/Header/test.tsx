import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Header, { HeaderProps } from '.';

jest.mock('./google-analytics', () => ({
  GoogleAnalytics: () => <div data-testid="google-analytics" />,
}));

jest.mock('./nav-item', () => {
  const MockNavItem = ({ label, href }: { label: string; href: string }) => (
    <a href={href} data-testid={`nav-item-${label}`}>
      {label}
    </a>
  );

  // Definindo a displayName
  MockNavItem.displayName = 'NavItem';

  return MockNavItem;
});

describe('Header component', () => {
  const defaultNavItems = [
    { label: 'Home', href: '/' },
    { label: 'Projects', href: '/projects' },
  ];

  it('renders the logo', () => {
    const { container } = render(<Header />);
    const logo = screen.getByAltText('Logo');
    expect(logo).toBeInTheDocument();
    expect(logo).toHaveAttribute('src', '/images/logo.svg');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the Google Analytics component', () => {
    render(<Header />);
    const googleAnalytics = screen.getByTestId('google-analytics');
    expect(googleAnalytics).toBeInTheDocument();
  });

  it('renders the navigation items by default', () => {
    render(<Header />);
    defaultNavItems.forEach(({ label }) => {
      const navItem = screen.getByTestId(`nav-item-${label}`);
      expect(navItem).toBeInTheDocument();
      expect(navItem).toHaveTextContent(label);
    });
  });

  it('renders custom navigation items when provided', () => {
    const customNavItems: HeaderProps['navItems'] = [
      { label: 'About', href: '/about' },
      { label: 'Contact', href: '/contact' },
    ];
    render(<Header navItems={customNavItems} />);
    customNavItems.forEach(({ label }) => {
      const navItem = screen.getByTestId(`nav-item-${label}`);
      expect(navItem).toBeInTheDocument();
      expect(navItem).toHaveTextContent(label);
    });
  });

  it('does not render default navigation items when custom items are provided', () => {
    const customNavItems: HeaderProps['navItems'] = [
      { label: 'About', href: '/about' },
    ];
    render(<Header navItems={customNavItems} />);
    defaultNavItems.forEach(({ label }) => {
      const navItem = screen.queryByTestId(`nav-item-${label}`);
      expect(navItem).not.toBeInTheDocument();
    });
  });
});
