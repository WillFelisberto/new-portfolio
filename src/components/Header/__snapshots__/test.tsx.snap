// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders the logo 1`] = `
<div
  data-testid="google-analytics"
/>
`;
